### ISSUE TITLE
KubOS CLI Ontology

### ISSUE THREAD

**Issue by damons** [kubos/kubos#341]
For the CLI, it seems to make sense to have some sort of command ontology that we can agree on.  This issue is to propose and track the words that are used.   As a first pass, I'll use what has already been [proposed](https://github.com/kubos/kubos/issues/334#issuecomment-486821904):

Top level command:  kubos

e.g.,: ` $ kubos  {args} `

And to be specific about the hierarchy:

`
kubos
-- install
---- config
-- start
---- ?   %  Any arguments, config options or sub-nouns or verbs?
-- stop
---- ?
-- create
---- ?
-- app
---- ?
`

Suggestions?

**Comment by marshall** [comment]
@damons yeah, this is a good idea. let's put this in a Markdown file in a gist or in a repo so we collaborate a bit more directly on this instead of commenting

**Comment by damons** [comment]
Will do.  Working on that now.

**Comment by jacoffey3** [comment]
@damons Tag me when it's up. I have some fills to put in.

**Comment by damons** [comment]
Please feel free to go ahead and add yours to whatever works (a markdown in a gist as Marshall recommended, or something).  I feel like I'd just be slowing people down if they waited on me to get this in some form that worked best for everyone.

**Comment by jacoffey3** [comment]
Top level command: kubos

e.g.,: $ kubos {args}

And to be specific about the hierarchy:

kubos

- `install`: makes sure Rust and Python are installed and available, throwing a warning if not, downloads and installs the packages for KubOS (app api, kubos service library, comm service library, etc)
  - `-version`: Release version, defaults to latest
- `config`
  - `-build`: Takes you through each service that's available and creates a config.toml for you to then edit as needed. Where does it put it? a `.kubos` directory in home? 
- `start`
  - `-config`: path to config toml, defaults to the location where build puts it by default (if that's not current working directory)
  - `-all`: all the services in the toml?
  - `-core`: only the kubOS Core services?
  - `-app`: only the mission app service? 
  - Not sure on the subsets, but they could be useful
- `stop`: Stops all running services
- `app`: example mission apps
  - `-create`: Drops an example in your working directory to help you get started
    - `{language}`: Python, Rust, (C in the future) 
  - `-install`: Installs into the local mission application service
  - `-run`: Installs into the mission application service and runs the application?
- `service`
  - `-create`: Drops an example in your working directory to help you get started, and adds the basic config requirements to your config.toml
    - `{language}`: Python, Rust
  - `-install`? Not really relevant locally since it just runs on it's own, but I'm not sure
  - `-run`?
- `version`: Release version of the services installed and version of the CLI

**Comment by jacoffey3** [comment]
To run each service, you need to run this command in the service's base folder: 
`$ cargo run -- -c /path/to/config.toml`

**Comment by plauche** [comment]
Hooray! Overall I think this structure goes a long way in creating a usable workflow that is inline with our architecture :smile:.

A few thoughts:
- `install` 
  - Does this also download/build/install (where?) copies of the middleware services?
- `config`
  - `~/.kubos` seems like an appropriate place for the `config.toml` as well as anything that is downloaded and/or built (services, libraries, apis, etc)
- `start`
  - Could also take `-service` parameter with the name of a service to start
- `stop`
  - `-service` parameter with the name of a service to stop
- `app`
  - I think under this command (or another?) we could expose all of the `app-service` functionality
- `query` - Gives the user the ability to easily query any running service

**Comment by CatGarab** [comment]
I'd prefer it if `kubos start` just started running whatever was in the current directory (for example, I'm developing an app. I navigate to `{path}/my-app-dir` and run `kubos start`. My app is then built and run)
- The "run whatever i'm working on" case will be the most common command run, so should be one of the shortest.
- I propose something like `kubos system start` for kicking off all the services

I don't know that I'd want to install an application every time I want to run it. I'd prefer to just run it locally (without installation) until I'm ready for final testing
- The app service expects a unique version number. This means we either have to go through a continuous cycle of uninstalling/installing for each change made to the app, or we auto-increment the app causing the app registry storage used to continuously expand

**Comment by CatGarab** [comment]
Wishlist: `kubos check`
- Rust - Wrapper around `cargo check`
- Python - Whatever static typing exists for Python now
- Validate the project's `manifest.toml` file

**Comment by damons** [comment]
@CatGarab Would 'check' also run tests?  Just curious.  Might be interesting.

**Comment by jacoffey3** [comment]
@damons We've had some developments, check it out: https://github.com/kubos/kubos/issues/352
